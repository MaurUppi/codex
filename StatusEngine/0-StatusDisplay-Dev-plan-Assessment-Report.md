---
stage: assessment
generated_at: 2025-09-09T14:12:21Z
accomplishment_path: /Users/ouzy/Documents/DevProjects/codex/project/0-StatusDisplay-Dev-plan-Accomplishment-report.md
dev_plan_path: project/0-StatusDisplay-Dev-plan.md
requirement_path: project/0-StatusDisplay-requirement.md
working_dir: /Users/ouzy/Documents/DevProjects/codex
branch: feat/statusengine
commit: 7176080
mode: reassessment
improvement_round: 3
findings_count: {critical: 0, medium: 1, low: 1}
previous_counts: {critical: 1, medium: 2, low: 0}
---

# Status Display & Timing — Assessment Report

## Verdict Summary
- Feature status: Complete
- Security posture: Strong
- Code quality: High
- Test coverage: Adequate (engine/helpers); Missing (TUI snapshots/integration)
- Documentation: Adequate
- Overall rating: 9/10
- Reassessment: Yes (Round 3)
- Delta since last assessment: Critical 1→0, Medium 2→1, Low 0→1

## Detailed Assessment

Context:
- Code improvements are on branch feat/statusengine.
- Development followed plan project/0-StatusDisplay-Dev-plan.md and requirement project/0-StatusDisplay-requirement.md.
- Accomplishment report path is /Users/ouzy/Documents/DevProjects/codex/project/0-StatusDisplay-Dev-plan-Accomplishment-report.md.

### Alignment to Plan & Requirements
- M1 (protocol timing): Optional since_session_ms field remains and is propagated across event emissions (per Round 1); no regressions observed.
- M2 (StatusEngine): Core complete with clamping, backoff/jitter, env sanitization, and tests.
- M3 (Git helpers): Best-effort counts; minor heuristic limitations remain acceptable.
- M4 (TUI integration): Now complete — engine wired into App state, tick cadence, footer rendering (Lines 2/3), desired height updated.
- M5 (Validation/rollout): Engine tests solid; TUI snapshot/integration tests still missing.

### Code Quality Analysis (evidence)

Protocol timing field (unchanged; correct and optional):
```rust path=/Users/ouzy/Documents/DevProjects/codex/codex-rs/protocol/src/protocol.rs start=398
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Event {
    pub id: String,
    pub msg: EventMsg,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub since_session_ms: Option<u64>,
}
```

App state wiring and tick cadence:
```rust path=/Users/ouzy/Documents/DevProjects/codex/codex-rs/tui/src/app.rs start=125
// Initialize StatusEngine if enabled with proper config mapping
let status_engine = if config.tui.statusengine.unwrap_or(false) {
    let status_config = Self::build_statusengine_config(&config.tui);
    let mut engine = StatusEngine::new(status_config);
    engine.set_line2_selection(&[
        StatusItem::Model,
        StatusItem::Effort,
        StatusItem::WorkspaceName,
        StatusItem::GitBranch,
        StatusItem::Sandbox,
        StatusItem::Approval,
    ]);
    Some(engine)
} else { None };
...
// Tick every 300ms and forward output to widget
if let Some(status_engine) = &mut app.status_engine {
    let now = Instant::now();
    let output = status_engine.tick(now).await;
    app.chat_widget.set_statusengine_output(Some(output));
    tui.frame_requester().schedule_frame();
}
```

Footer layout and desired height with Lines 2/3:
```rust path=/Users/ouzy/Documents/DevProjects/codex/codex-rs/tui/src/bottom_pane/chat_composer.rs start=158
pub fn desired_height(&self, width: u16) -> u16 {
    self.textarea.desired_height(width - 1)
        + match &self.active_popup {
            ActivePopup::None => {
                if self.statusengine_enabled { 3u16 } else { 1u16 }
            }
            ActivePopup::Command(c) => c.calculate_required_height(),
            ActivePopup::File(c) => c.calculate_required_height(),
        }
}
```

Rendering Lines 2/3 from StatusEngine output:
```rust path=/Users/ouzy/Documents/DevProjects/codex/codex-rs/tui/src/bottom_pane/chat_composer.rs start=1389
fn render_statusengine_lines(&self, area: Rect, buf: &mut Buffer) {
    if area.height >= 2 {
        let [line2_rect, line3_rect] = Layout::vertical([Constraint::Length(1), Constraint::Length(1)]).areas(area);
        if let Some(ref output) = self.statusengine_output {
            Line::from(output.line2.clone()).style(Style::default().dim()).render_ref(line2_rect, buf);
            if let Some(ref line3) = output.line3 {
                Line::from(line3.clone()).style(Style::default().dim()).render_ref(line3_rect, buf);
            }
        }
    }
}
```

### Security Review
- Provider subprocess hardening: env cleared with minimal PATH, kill_on_drop, explicit kill on timeout; throttle with jitter and exponential backoff — strong posture.
- Git operations bounded by timeouts — safe against UI stalls.
- No sensitive data present in payloads beyond paths and state metadata.

### Test Coverage Analysis
- Engine: unit tests present and organized under codex-rs/tui/tests.
- Git helpers: robust temporary-repo tests.
- TUI: still lacks snapshot or end-to-end tests validating footer lines and width behavior.

### Behavioral Outcomes
- TUI now displays Line 2 and optional Line 3; height adjusts accordingly; updates flow within a frame after state changes.
- No regressions observed in protocol or engine behavior since Round 2.

## Findings & Recommendations (ordered)

### Critical
- None.

### Medium
1) Missing TUI snapshot/integration tests
- Impact: Footer rendering and width/ellipsis behavior not guarded; potential regressions.
- Recommendations:
  - Add insta-based snapshots for narrow/medium/wide widths under codex-rs/tui/tests/snapshots/.
  - Add an integration test to simulate StatusEngine outputs in ChatComposer and assert desired_height and rendering lines.

### Low
1) Provider environment allowlisting and docs
- Impact: While env_clear is used with PATH restored, consider explicit allowlisting if future needs arise (e.g., locale variables) and document the policy.
- Recommendations:
  - Document env semantics in README/TUI docs and ensure future additions go through review.

## Acceptance Criteria Check
- Timing: since_session_ms present and non-decreasing (logical path intact) — satisfied.
- StatusEngine: Line 2 composition and Line 3 provider with 300 ms throttle and timeouts — satisfied.
- Rendering: Footer Lines 2/3 integrated with height adjustments — satisfied.
- Config: Mapping present in app initialization with clamping/validation — satisfied.

## Conclusion
Round 3 completes the user-visible integration. The engine is robust and secured; Git helpers are safe; TUI wiring and rendering are in place. Remaining work is limited to test coverage for rendering.

Overall: Feature ready; maintain with added snapshots and minor docs.

## Appendices
- Key diffs/paths reviewed (since last assessment):
  - Commits: 25c561c (Round 3 integration), 7176080 (formatting)
  - Files touched in Round 3: codex-rs/tui/src/app.rs; codex-rs/tui/src/bottom_pane/{chat_composer.rs,mod.rs}; codex-rs/tui/src/chatwidget.rs; codex-rs/core/src/git_info.rs
  - Merge-base vs origin/main: 62bd0e3

