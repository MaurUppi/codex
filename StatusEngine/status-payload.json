{"session_id":"965f51fd-c523-4369-bf4a-5f832cf1aafa","model":{"id":"gpt-5"},"effort":"medium","workspace":{"name":"codex","current_dir":"/Users/ouzy/Documents/DevProjects/codex"},"cwd":"/Users/ouzy/Documents/DevProjects/codex","git":{"branch":"feat/statusengine","counts":"+1 -1"},"sandbox":"workspace-write","approval":"on-request","timing":{"since_session_ms":189752}}