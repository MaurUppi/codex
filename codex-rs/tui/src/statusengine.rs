//! StatusEngine - Manages TUI footer status display with timing, git info, and external providers.

use ratatui::style::Stylize;
use serde_json;
use std::path::PathBuf;
use std::process::Stdio;
use std::time::Duration;
use std::time::Instant;
use tokio::io::AsyncWriteExt;
use tokio::process::Command;
use tokio::time::timeout;
use tracing;

// Import git helpers for branch and diff count information
use codex_core::git_info::{collect_git_info, working_diff_counts};

/// Configuration for the StatusEngine
#[derive(Debug, Clone)]
pub struct StatusEngineConfig {
    /// Timeout for external command provider in milliseconds
    pub command_timeout_ms: u64,
    /// Optional path to external command provider
    pub command: Option<String>,
    /// Whether StatusEngine is enabled
    pub enabled: bool,
    /// Provider type: "command" or "builtin"
    pub provider: String,
}

impl Default for StatusEngineConfig {
    fn default() -> Self {
        Self {
            command_timeout_ms: 300,
            command: None,
            enabled: false,
            provider: "builtin".to_string(),
        }
    }
}

/// Available status items for Line 2
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StatusItem {
    Model,
    Effort,
    WorkspaceName,
    GitBranch,
    Sandbox,
    Approval,
}

/// Current state of the session for status display
#[derive(Debug, Clone, Default)]
pub struct StatusEngineState {
    pub model: Option<String>,
    pub effort: Option<String>,
    pub workspace_name: Option<String>,
    pub git_branch: Option<String>,
    pub git_counts: Option<String>,
    pub sandbox: Option<String>,
    pub approval: Option<String>,
    pub since_session_ms: Option<u64>,
    pub cwd: Option<PathBuf>,
}

/// Output from the StatusEngine
#[derive(Debug, Clone, Default)]
pub struct StatusEngineOutput {
    pub line2: String,
    pub line3: Option<String>,
}

/// Main StatusEngine implementation
pub struct StatusEngine {
    config: StatusEngineConfig,
    state: StatusEngineState,
    line2_items: Vec<StatusItem>,
    last_command_run: Option<Instant>,
    last_line3: Option<String>,
    command_cooldown: Duration,
    consecutive_failures: u32,
    backoff_until: Option<Instant>,
}

impl StatusEngine {
    /// Create a new StatusEngine with the given configuration
    pub fn new(mut config: StatusEngineConfig) -> Self {
        // Validate and clamp configuration values

        // Clamp command timeout to reasonable range (150-500ms as per assessment)
        config.command_timeout_ms = config.command_timeout_ms.clamp(150, 500);

        // Validate provider type with fallback to "builtin"
        if config.provider != "command" && config.provider != "builtin" {
            tracing::debug!(
                "StatusEngine: invalid provider '{}', falling back to 'builtin'",
                config.provider
            );
            config.provider = "builtin".to_string();
        }

        // Default order from the requirement
        let default_items = vec![
            StatusItem::Model,
            StatusItem::Effort,
            StatusItem::WorkspaceName,
            StatusItem::GitBranch,
            StatusItem::Sandbox,
            StatusItem::Approval,
        ];

        Self {
            config,
            state: StatusEngineState::default(),
            line2_items: default_items,
            last_command_run: None,
            last_line3: None,
            command_cooldown: Duration::from_millis(300), // Built-in 300ms throttle
            consecutive_failures: 0,
            backoff_until: None,
        }
    }

    /// Update the engine state with new session information
    pub fn set_state(&mut self, state: StatusEngineState) {
        self.state = state;
    }

    /// Set the items and order for Line 2 display
    pub fn set_line2_selection(&mut self, items: &[StatusItem]) {
        self.line2_items = items.to_vec();
    }

    /// Get the current Line 2 items (for testing)
    pub fn line2_items(&self) -> &[StatusItem] {
        &self.line2_items
    }

    /// Apply consistent styling to status line text
    fn style_status_line(text: String) -> String {
        text.dim().to_string()
    }

    /// Tick the engine and produce status output
    /// Respects the 300ms throttle for external provider calls
    pub async fn tick(&mut self, now: Instant) -> StatusEngineOutput {
        // Update git information before building Line 2
        self.update_git_info().await;
        
        let line2 = self.build_line2();
        let line3 = self.maybe_run_command_provider(now).await;

        StatusEngineOutput { line2, line3 }
    }

    /// Update git branch and diff counts information
    /// This is called on each tick to refresh git status
    async fn update_git_info(&mut self) {
        if let Some(ref cwd) = self.state.cwd {
            // Get current git branch from git info
            if let Some(git_info) = collect_git_info(cwd).await {
                self.state.git_branch = git_info.branch;
            }
            
            // Get diff counts (+added, -removed) against HEAD
            if let Some((added, removed)) = working_diff_counts(cwd).await {
                self.state.git_counts = Some(format!("+{} -{}", added, removed));
            }
        }
    }

    /// Build Line 2 from selected status items
    /// Made public for testing purposes
    pub fn build_line2(&self) -> String {
        let mut parts = Vec::new();

        for item in &self.line2_items {
            match item {
                StatusItem::Model => {
                    if let Some(ref model) = self.state.model {
                        parts.push(model.clone());
                    }
                }
                StatusItem::Effort => {
                    if let Some(ref effort) = self.state.effort {
                        parts.push(effort.clone());
                    }
                }
                StatusItem::WorkspaceName => {
                    if let Some(ref name) = self.state.workspace_name {
                        parts.push(name.clone());
                    }
                }
                StatusItem::GitBranch => {
                    if let Some(ref branch) = self.state.git_branch {
                        let git_part = if let Some(ref counts) = self.state.git_counts {
                            format!("{branch} {counts}")
                        } else {
                            branch.clone()
                        };
                        parts.push(git_part);
                    }
                }
                StatusItem::Sandbox => {
                    if let Some(ref sandbox) = self.state.sandbox {
                        parts.push(sandbox.clone());
                    }
                }
                StatusItem::Approval => {
                    if let Some(ref approval) = self.state.approval {
                        parts.push(approval.clone());
                    }
                }
            }
        }

        // Join with " | " separator and apply consistent styling
        if parts.is_empty() {
            String::new()
        } else {
            Self::style_status_line(parts.join(" | "))
        }
    }

    /// Check if we should run the command provider and execute if so
    /// Made public for testing purposes
    pub async fn maybe_run_command_provider(&mut self, now: Instant) -> Option<String> {
        // Only run if provider is "command" and command is configured
        if self.config.provider != "command" || self.config.command.is_none() {
            return None;
        }

        // Check if we're in backoff period
        if let Some(backoff_until) = self.backoff_until {
            if now < backoff_until {
                tracing::debug!(
                    "StatusEngine command provider in backoff period, using cached result"
                );
                return self.last_line3.clone();
            } else {
                // Backoff period expired, clear it
                tracing::debug!(
                    "StatusEngine backoff period expired, resuming command provider calls"
                );
                self.backoff_until = None;
            }
        }

        // Check throttling with jitter
        let jitter_ms = (now.elapsed().as_nanos() % 100) as u64; // Simple jitter 0-99ms
        let effective_cooldown = self.command_cooldown + Duration::from_millis(jitter_ms);

        if let Some(last_run) = self.last_command_run
            && now.duration_since(last_run) < effective_cooldown
        {
            tracing::debug!("StatusEngine command provider throttled, using cached result");
            return self.last_line3.clone();
        }

        // Run the command
        match self.run_command_provider().await {
            Ok(Some(output)) => {
                tracing::debug!("StatusEngine command provider succeeded, got output");
                self.last_command_run = Some(now);
                self.last_line3 = Some(output.clone());
                self.consecutive_failures = 0; // Reset failure count on success
                Some(output)
            }
            Ok(None) => {
                tracing::debug!("StatusEngine command provider returned empty result");
                self.last_command_run = Some(now);
                self.consecutive_failures = 0; // Empty result is not a failure
                // Keep last good output on empty result
                self.last_line3.clone()
            }
            Err(e) => {
                tracing::debug!("StatusEngine command provider error: {}", e);
                self.last_command_run = Some(now);
                self.consecutive_failures += 1;

                // Apply exponential backoff after failures
                if self.consecutive_failures >= 3 {
                    let backoff_ms =
                        std::cmp::min(5000, 1000 * (1 << (self.consecutive_failures - 3)));
                    self.backoff_until = Some(now + Duration::from_millis(backoff_ms));
                    tracing::debug!(
                        "StatusEngine entering backoff mode for {}ms after {} consecutive failures",
                        backoff_ms,
                        self.consecutive_failures
                    );
                }

                // Keep last good output on error
                self.last_line3.clone()
            }
        }
    }

    /// Execute the configured command provider
    async fn run_command_provider(
        &self,
    ) -> Result<Option<String>, Box<dyn std::error::Error + Send + Sync>> {
        let command_path = match &self.config.command {
            Some(cmd) => cmd,
            None => return Ok(None),
        };

        // Build JSON payload
        let payload = self.build_command_payload()?;
        let payload_json = serde_json::to_string(&payload)?;

        // Spawn the command with timeout and proper cleanup
        let timeout_duration = Duration::from_millis(self.config.command_timeout_ms);
        tracing::debug!(
            "StatusEngine spawning command provider with timeout {}ms",
            self.config.command_timeout_ms
        );

        let mut child = Command::new(command_path)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .kill_on_drop(true) // Ensure child is killed if dropped
            .env_clear() // Clear environment for security
            .env("PATH", std::env::var("PATH").unwrap_or_default()) // Keep minimal PATH
            .spawn()?;

        let result = timeout(timeout_duration, async move {
            // Write payload to stdin
            if let Some(stdin) = child.stdin.as_mut() {
                stdin.write_all(payload_json.as_bytes()).await?;
                stdin.shutdown().await?;
            }

            // Wait for completion and get output
            let output = child.wait_with_output().await?;

            if output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);
                // Get first line only
                let first_line = stdout.lines().next().unwrap_or("").trim().to_string();
                if first_line.is_empty() {
                    Ok::<Option<String>, Box<dyn std::error::Error + Send + Sync>>(None)
                } else {
                    Ok(Some(first_line))
                }
            } else {
                Ok(None)
            }
        })
        .await;

        match result {
            Ok(Ok(output)) => Ok(output),
            Ok(Err(e)) => Err(e),
            Err(_) => {
                // Timeout occurred - child process should be killed by kill_on_drop(true)
                tracing::debug!(
                    "StatusEngine command provider timed out, child will be killed on drop"
                );
                Ok(None) // Return None to keep last good output
            }
        }
    }

    /// Build JSON payload for command provider
    fn build_command_payload(&self) -> Result<serde_json::Value, serde_json::Error> {
        let mut payload = serde_json::Map::new();

        // Add available fields
        if let Some(ref model) = self.state.model {
            let mut model_obj = serde_json::Map::new();
            model_obj.insert("id".to_string(), serde_json::Value::String(model.clone()));
            payload.insert("model".to_string(), serde_json::Value::Object(model_obj));
        }

        if let Some(ref effort) = self.state.effort {
            payload.insert(
                "effort".to_string(),
                serde_json::Value::String(effort.clone()),
            );
        }

        if let Some(ref workspace_name) = self.state.workspace_name {
            let mut workspace_obj = serde_json::Map::new();
            workspace_obj.insert(
                "name".to_string(),
                serde_json::Value::String(workspace_name.clone()),
            );
            if let Some(ref cwd) = self.state.cwd {
                workspace_obj.insert(
                    "current_dir".to_string(),
                    serde_json::Value::String(cwd.display().to_string()),
                );
                // Only add project_dir if it differs from current_dir
                // In the current implementation they're the same, so we omit project_dir
                // to avoid duplication as recommended in the assessment
            }
            payload.insert(
                "workspace".to_string(),
                serde_json::Value::Object(workspace_obj),
            );
        }

        if let Some(ref cwd) = self.state.cwd {
            payload.insert(
                "cwd".to_string(),
                serde_json::Value::String(cwd.display().to_string()),
            );
        }

        // Git info
        if self.state.git_branch.is_some() || self.state.git_counts.is_some() {
            let mut git_obj = serde_json::Map::new();
            if let Some(ref branch) = self.state.git_branch {
                git_obj.insert(
                    "branch".to_string(),
                    serde_json::Value::String(branch.clone()),
                );
            }
            if let Some(ref counts) = self.state.git_counts {
                git_obj.insert(
                    "counts".to_string(),
                    serde_json::Value::String(counts.clone()),
                );
            }
            payload.insert("git".to_string(), serde_json::Value::Object(git_obj));
        }

        if let Some(ref sandbox) = self.state.sandbox {
            payload.insert(
                "sandbox".to_string(),
                serde_json::Value::String(sandbox.clone()),
            );
        }

        if let Some(ref approval) = self.state.approval {
            payload.insert(
                "approval".to_string(),
                serde_json::Value::String(approval.clone()),
            );
        }

        // Timing info
        if let Some(since_session_ms) = self.state.since_session_ms {
            let mut timing_obj = serde_json::Map::new();
            timing_obj.insert(
                "since_session_ms".to_string(),
                serde_json::Value::Number(since_session_ms.into()),
            );
            payload.insert("timing".to_string(), serde_json::Value::Object(timing_obj));
        }

        Ok(serde_json::Value::Object(payload))
    }

    /// Apply center-ellipsis truncation to a string for the given width
    /// Made public for testing purposes
    pub fn truncate_with_ellipsis(text: &str, max_width: usize) -> String {
        if text.len() <= max_width {
            return text.to_string();
        }

        if max_width < 3 {
            return "...".chars().take(max_width).collect();
        }

        let ellipsis = "…";
        let available = max_width - ellipsis.len();
        let left_len = available / 2;
        let right_len = available - left_len;

        let chars: Vec<char> = text.chars().collect();
        let left_part: String = chars.iter().take(left_len).collect();
        let right_part: String = chars
            .iter()
            .rev()
            .take(right_len)
            .collect::<String>()
            .chars()
            .rev()
            .collect();

        format!("{left_part}{ellipsis}{right_part}")
    }

    /// Get width-aware Line 2 with truncation applied specifically to branch token
    pub fn get_line2_with_width(&self, max_width: usize) -> String {
        let mut parts = Vec::new();

        for item in &self.line2_items {
            match item {
                StatusItem::Model => {
                    if let Some(ref model) = self.state.model {
                        parts.push(model.clone());
                    }
                }
                StatusItem::Effort => {
                    if let Some(ref effort) = self.state.effort {
                        parts.push(effort.clone());
                    }
                }
                StatusItem::WorkspaceName => {
                    if let Some(ref name) = self.state.workspace_name {
                        parts.push(name.clone());
                    }
                }
                StatusItem::GitBranch => {
                    if let Some(ref branch) = self.state.git_branch {
                        let mut git_part = if let Some(ref counts) = self.state.git_counts {
                            format!("{branch} {counts}")
                        } else {
                            branch.clone()
                        };

                        // Calculate available width for branch token
                        let separator_len = if parts.is_empty() { 0 } else { " | ".len() };
                        let other_parts_len: usize = parts.iter().map(|p| p.len()).sum::<usize>()
                            + (parts.len().saturating_sub(1)) * " | ".len(); // separators between existing parts
                        let remaining_parts_estimate =
                            (self.line2_items.len() - parts.len() - 1) * 15; // estimate for remaining
                        let available_for_branch = max_width.saturating_sub(
                            other_parts_len + separator_len + remaining_parts_estimate,
                        );

                        if git_part.len() > available_for_branch && available_for_branch > 3 {
                            git_part =
                                Self::truncate_with_ellipsis(&git_part, available_for_branch);
                        }
                        parts.push(git_part);
                    }
                }
                StatusItem::Sandbox => {
                    if let Some(ref sandbox) = self.state.sandbox {
                        parts.push(sandbox.clone());
                    }
                }
                StatusItem::Approval => {
                    if let Some(ref approval) = self.state.approval {
                        parts.push(approval.clone());
                    }
                }
            }
        }

        // Join with " | " separator and apply dim styling
        if parts.is_empty() {
            String::new()
        } else {
            let result = parts.join(" | ");
            // If still too long after branch truncation, truncate the whole line as fallback
            if result.len() > max_width {
                Self::style_status_line(Self::truncate_with_ellipsis(&result, max_width))
            } else {
                Self::style_status_line(result)
            }
        }
    }
}
