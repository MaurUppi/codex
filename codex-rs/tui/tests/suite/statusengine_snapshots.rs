//! StatusEngine TUI snapshot and integration tests
//!
//! This module tests footer rendering with StatusEngine output at different widths,
//! ensuring the layout behaves correctly and text truncation works as expected.

use codex_tui::app_event_sender::AppEventSender;
use codex_tui::bottom_pane::chat_composer::ChatComposer;
use codex_tui::statusengine::StatusEngine;
use codex_tui::statusengine::StatusEngineConfig;
use codex_tui::statusengine::StatusEngineOutput;
use codex_tui::statusengine::StatusEngineState;
use codex_tui::statusengine::StatusItem;
use ratatui::Terminal;
use ratatui::backend::TestBackend;
use ratatui::layout::Rect;
use ratatui::widgets::WidgetRef;
use tokio::sync::mpsc;

/// Test helper to create a ChatComposer with StatusEngine enabled
fn create_test_chat_composer(statusengine_output: Option<StatusEngineOutput>) -> ChatComposer {
    let (tx, _rx) = mpsc::unbounded_channel();
    let app_event_tx = AppEventSender::new(tx);

    let mut composer = ChatComposer::new(
        false, // has_input_focus
        app_event_tx,
        true,          // enhanced_keys_supported
        String::new(), // placeholder_text
        false,         // disable_paste_burst
        true,          // statusengine_enabled
    );

    // Set StatusEngine output
    composer.set_statusengine_output(statusengine_output);

    composer
}

/// Create a test StatusEngineOutput with sample data
fn create_sample_output(line2: &str, line3: Option<&str>) -> StatusEngineOutput {
    StatusEngineOutput {
        line2: line2.to_string(),
        line3: line3.map(|s| s.to_string()),
    }
}

/// Test helper to render StatusEngine footer at a specific width
fn render_statusengine_footer(
    output: Option<StatusEngineOutput>,
    width: u16,
    height: u16,
) -> String {
    let composer = create_test_chat_composer(output);
    let backend = TestBackend::new(width, height);
    let mut terminal = Terminal::new(backend).unwrap_or_else(|e| {
        panic!("Failed to create test terminal for StatusEngine rendering: {e}")
    });

    terminal
        .draw(|f| {
            let area = Rect::new(0, 0, width, height);
            composer.render_ref(area, f.buffer_mut());
        })
        .unwrap_or_else(|e| panic!("Failed to draw StatusEngine footer to test terminal: {e}"));

    terminal
        .backend()
        .buffer()
        .clone()
        .content()
        .iter()
        .map(|cell| cell.symbol())
        .collect::<String>()
}

#[test]
fn test_desired_height_without_statusengine() {
    let (tx, _rx) = mpsc::unbounded_channel();
    let app_event_tx = AppEventSender::new(tx);

    let composer = ChatComposer::new(
        false, // has_input_focus
        app_event_tx,
        true,          // enhanced_keys_supported
        String::new(), // placeholder_text
        false,         // disable_paste_burst
        false,         // statusengine_enabled (disabled)
    );

    // Expected: 1 line for textarea + 1 line for hints = 2
    assert_eq!(composer.desired_height(80), 2);
}

#[test]
fn test_desired_height_with_statusengine() {
    let output = create_sample_output("model: gpt-4 | effort: auto", Some("git: main (+2 -1)"));
    let composer = create_test_chat_composer(Some(output));

    // Expected: 1 line for textarea + (hints + line2 + line3) = 1 + 3 = 4
    assert_eq!(composer.desired_height(80), 4);
}

#[test]
fn test_statusengine_lines_rendering_narrow_width() {
    let output = create_sample_output(
        "model: claude-sonnet-3.5 | effort: auto | workspace: codex | git: feat/statusengine (+15 -2) | sandbox: read-only",
        Some("provider: git status --porcelain"),
    );

    let result = render_statusengine_footer(Some(output), 40, 3);
    insta::assert_snapshot!("statusengine_narrow_40", result);
}

#[test]
fn test_statusengine_lines_rendering_medium_width() {
    let output = create_sample_output(
        "model: claude-sonnet-3.5 | effort: auto | workspace: codex | git: feat/statusengine (+15 -2) | sandbox: read-only | approval: on-request",
        Some("provider: git status --porcelain && git log --oneline -1"),
    );

    let result = render_statusengine_footer(Some(output), 80, 3);
    insta::assert_snapshot!("statusengine_medium_80", result);
}

#[test]
fn test_statusengine_lines_rendering_wide_width() {
    let output = create_sample_output(
        "model: claude-sonnet-3.5 | effort: auto | workspace: codex-development-environment | git: feat/statusengine-integration (+25 -8) | sandbox: workspace-write | approval: on-request",
        Some(
            "provider: git status --porcelain --branch && git log --oneline -3 && git diff --stat HEAD~1",
        ),
    );

    let result = render_statusengine_footer(Some(output), 120, 3);
    insta::assert_snapshot!("statusengine_wide_120", result);
}

#[test]
fn test_statusengine_line2_only() {
    let output = create_sample_output(
        "model: gpt-4 | effort: auto | workspace: test",
        None, // No line3
    );

    let result = render_statusengine_footer(Some(output), 80, 3);
    insta::assert_snapshot!("statusengine_line2_only", result);
}

#[test]
fn test_statusengine_disabled() {
    let result = render_statusengine_footer(None, 80, 1);
    insta::assert_snapshot!("statusengine_disabled", result);
}

#[test]
fn test_statusengine_truncation_behavior() {
    // Test various scenarios where text needs truncation
    let long_output = create_sample_output(
        "model: claude-3-5-sonnet-20241022 | effort: automatically-determined | workspace: very-long-workspace-name-that-should-truncate | git: feature/very-long-branch-name-with-lots-of-changes (+999 -888) | sandbox: danger-full-access | approval: on-request-with-confirmation",
        Some(
            "provider: git status --porcelain --branch && git log --oneline -10 && git diff --stat HEAD~5 && git branch -av",
        ),
    );

    // Test at different widths to verify truncation
    let narrow = render_statusengine_footer(Some(long_output.clone()), 30, 3);
    insta::assert_snapshot!("statusengine_truncation_narrow_30", narrow);

    let medium = render_statusengine_footer(Some(long_output.clone()), 60, 3);
    insta::assert_snapshot!("statusengine_truncation_medium_60", medium);

    let wide = render_statusengine_footer(Some(long_output), 100, 3);
    insta::assert_snapshot!("statusengine_truncation_wide_100", wide);
}

/// Integration test: simulate StatusEngine tick and verify ChatComposer integration
#[tokio::test]
async fn test_statusengine_chatcomposer_integration() {
    // Create a StatusEngine with realistic configuration
    let config = StatusEngineConfig {
        provider: "builtin".to_string(),
        command: None,
        command_timeout_ms: 300,
        enabled: false,
    };

    let mut engine = StatusEngine::new(config);
    engine.set_line2_selection(&[
        StatusItem::Model,
        StatusItem::Effort,
        StatusItem::WorkspaceName,
        StatusItem::GitBranch,
        StatusItem::Sandbox,
        StatusItem::Approval,
    ]);

    // Set realistic state
    let state = StatusEngineState {
        model: Some("gpt5".to_string()),
        effort: Some("auto".to_string()),
        workspace_name: Some("codex".to_string()),
        git_branch: Some("feat/statusengine".to_string()),
        git_counts: Some("+5 -2 ?1".to_string()), // staged, unstaged, untracked
        sandbox: Some("read-only".to_string()),
        approval: Some("on-request".to_string()),
        ..Default::default()
    };

    engine.set_state(state);

    // Tick the engine to get output
    let now = std::time::Instant::now();
    let output = engine.tick(now).await;

    // Verify the output structure
    assert!(
        !output.line2.is_empty(),
        "Line2 should contain status information"
    );
    assert!(output.line2.contains("gpt5"), "Should contain model name");
    assert!(
        output.line2.contains("feat/statusengine"),
        "Should contain branch name"
    );
    assert!(output.line2.contains("+5"), "Should contain git counts");

    // Test ChatComposer integration
    let composer = create_test_chat_composer(Some(output.clone()));

    // Test desired height calculation
    assert_eq!(
        composer.desired_height(80),
        4,
        "Height should be 4 with StatusEngine enabled (textarea + hints + 2 status lines)"
    );

    // Test rendering at different widths
    let narrow_render = render_statusengine_footer(Some(output.clone()), 40, 3);
    let wide_render = render_statusengine_footer(Some(output.clone()), 120, 3);

    // The narrow version should be different from wide (due to truncation)
    assert_ne!(
        narrow_render, wide_render,
        "Rendering should differ based on width"
    );

    // Both should render some content
    assert!(
        !narrow_render.trim().is_empty(),
        "Narrow render should not be empty"
    );
    assert!(
        !wide_render.trim().is_empty(),
        "Wide render should not be empty"
    );
}

/// Test ellipsis truncation behavior specifically
#[test]
fn test_statusengine_ellipsis_truncation() {
    let output = create_sample_output(
        "model: claude-3-5-sonnet-20241022-very-long-model-name | effort: automatically-determined-with-high-confidence | workspace: extremely-long-workspace-name-that-definitely-needs-truncation",
        None,
    );

    // Test at very narrow width where ellipsis should appear
    let result = render_statusengine_footer(Some(output), 25, 3);
    insta::assert_snapshot!("statusengine_ellipsis_25", result);
}
