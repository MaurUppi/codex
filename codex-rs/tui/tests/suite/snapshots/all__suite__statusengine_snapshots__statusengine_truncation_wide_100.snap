---
source: tui/tests/suite/statusengine_snapshots.rs
assertion_line: 172
expression: wide
---
▌                                                                                                    ⏎ send   ⇧⏎ newline   ⌃T transcript   ⌃C quit
